addons/crm/models/crm_lead.py

```diff
@@ -382,31 +388,56 @@ def create(self, vals_list):
         return leads
 
     def write(self, vals):
-        # stage change:
-        if 'stage_id' in vals:
-            stage_id = self.env['crm.stage'].browse(vals['stage_id'])
-            if stage_id.is_won:
-                vals.update({'probability': 100})
-
-        # stage change with new stage: update probability and date_closed
+        #   stage change with new stage: update probability and date_closed
         if vals.get('probability', 0) >= 100 or not vals.get('active', True):
             vals['date_closed'] = fields.Datetime.now()
         elif 'probability' in vals:
             vals['date_closed'] = False
 
+        if any(field in ['active', 'stage_id'] for field in vals):
+            vals = self._handle_won_lost(vals)
```

Comment:
I would separate the fact that you set the probability to 100% if new_stage is "won" and this method.
Seems like it's mixing vals update and it does too much stuff.

The probability thing was better when it was here above.